File: Retainic/AudioManager.swift
//
//  AudioManager.swift
//  Retainic
//
//  Records a pronunciation clip (AVAudioRecorder) and plays back local or
//  Firebase Storage audio (AVAudioPlayer).
//

import Foundation
import Combine
import AVFoundation

// MARK: - Recorder (used while adding/editing a word)

@MainActor
final class PronunciationRecorder: NSObject, ObservableObject {
    @Published var isRecording = false
    @Published var isPlaying = false
    @Published var permissionDenied = false
    /// Set when a brand-new clip has been recorded this session.
    @Published private(set) var recordedURL: URL?

    /// Storage path of an already-saved recording (when editing a word).
    @Published var existingAudioPath: String?

    private var recorder: AVAudioRecorder?
    private var player: AVAudioPlayer?

    /// Whether there is any audio to play / save (new or previously saved).
    var hasAudio: Bool { recordedURL != nil || existingAudioPath != nil }
    /// Whether the user recorded a new clip that needs uploading.
    var hasNewRecording: Bool { recordedURL != nil }

    func configure(existingAudioPath: String?) {
        self.existingAudioPath = existingAudioPath
    }

    // MARK: Recording

    func toggleRecording() {
        isRecording ? stopRecording() : requestAndStartRecording()
    }

    private func requestAndStartRecording() {
        AVAudioApplication.requestRecordPermission { [weak self] granted in
            Task { @MainActor in
                guard let self else { return }
                if granted {
                    self.beginRecording()
                } else {
                    self.permissionDenied = true
                }
            }
        }
    }

    private func beginRecording() {
        stopPlayback()
        do {
            let session = AVAudioSession.sharedInstance()
            try session.setCategory(.playAndRecord, mode: .default, options: [.defaultToSpeaker])
            try session.setActive(true)

            let url = FileManager.default.temporaryDirectory
                .appendingPathComponent("retainic-rec-\(UUID().uuidString).m4a")
            let settings: [String: Any] = [
                AVFormatIDKey: Int(kAudioFormatMPEG4AAC),
                AVSampleRateKey: 44_100,
                AVNumberOfChannelsKey: 1,
                AVEncoderAudioQualityKey: AVAudioQuality.high.rawValue,
            ]
            let recorder = try AVAudioRecorder(url: url, settings: settings)
            recorder.record()
            self.recorder = recorder
            self.recordedURL = url
            isRecording = true
        } catch {
            isRecording = false
        }
    }

    private func stopRecording() {
        recorder?.stop()
        recorder = nil
        isRecording = false
        try? AVAudioSession.sharedInstance().setActive(false)
    }

    /// Discards the current recording (new clip or reference to a saved one).
    func clear() {
        stopPlayback()
        stopRecording()
        if let url = recordedURL { try? FileManager.default.removeItem(at: url) }
        recordedURL = nil
        existingAudioPath = nil
    }

    // MARK: Playback

    func play() {
        if let url = recordedURL {
            play(localURL: url)
        } else if let path = existingAudioPath {
            Task { await playRemote(path: path) }
        }
    }

    private func playRemote(path: String) async {
        do {
            let data = try await VocabRepository.downloadAudioData(path: path)
            play(data: data)
        } catch {
            isPlaying = false
        }
    }

    private func play(localURL: URL) {
        guard let data = try? Data(contentsOf: localURL) else { return }
        play(data: data)
    }

    private func play(data: Data) {
        do {
            try AVAudioSession.sharedInstance().setCategory(.playback)
            try AVAudioSession.sharedInstance().setActive(true)
            let player = try AVAudioPlayer(data: data)
            player.delegate = self
            player.play()
            self.player = player
            isPlaying = true
        } catch {
            isPlaying = false
        }
    }

    func stopPlayback() {
        player?.stop()
        player = nil
        isPlaying = false
    }
}

extension PronunciationRecorder: AVAudioPlayerDelegate {
    nonisolated func audioPlayerDidFinishPlaying(_ player: AVAudioPlayer, successfully flag: Bool) {
        Task { @MainActor in self.isPlaying = false }
    }
}

// MARK: - Playback-only store (used in lists / flashcards)

/// Shared player for tapping a word's pronunciation outside the editor.
@MainActor
final class AudioPlaybackStore: NSObject, ObservableObject {
    static let shared = AudioPlaybackStore()

    /// Storage path currently playing, for highlighting the active button.
    @Published private(set) var playingPath: String?

    private var player: AVAudioPlayer?
    private var cache: [String: Data] = [:]

    func toggle(path: String) {
        if playingPath == path {
            stop()
        } else {
            Task { await play(path: path) }
        }
    }

    private func play(path: String) async {
        stop()
        do {
            let data: Data
            if let cached = cache[path] {
                data = cached
            } else {
                data = try await VocabRepository.downloadAudioData(path: path)
                cache[path] = data
            }
            try AVAudioSession.sharedInstance().setCategory(.playback)
            try AVAudioSession.sharedInstance().setActive(true)
            let player = try AVAudioPlayer(data: data)
            player.delegate = self
            player.play()
            self.player = player
            playingPath = path
        } catch {
            playingPath = nil
        }
    }

    func stop() {
        player?.stop()
        player = nil
        playingPath = nil
    }
}

extension AudioPlaybackStore: AVAudioPlayerDelegate {
    nonisolated func audioPlayerDidFinishPlaying(_ player: AVAudioPlayer, successfully flag: Bool) {
        Task { @MainActor in self.playingPath = nil }
    }
}

File: android/app/src/main/java/com/retainic/app/audio/AudioManager.kt
package com.retainic.app.audio

import android.content.Context
import android.media.MediaPlayer
import android.media.MediaRecorder
import android.os.Build
import android.speech.tts.TextToSpeech
import android.speech.tts.UtteranceProgressListener
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue
import com.retainic.app.data.VocabRepository
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import java.io.File
import java.util.Locale

/**
 * Records a pronunciation clip (MediaRecorder) and plays back local or Firebase
 * Storage audio (MediaPlayer). Ported from AudioManager.swift.
 */
class PronunciationRecorder(private val context: Context) {
    var isRecording by mutableStateOf(false)
        private set
    var isPlaying by mutableStateOf(false)
        private set
    var permissionDenied by mutableStateOf(false)
    /** Set when the last recording captured no audio. */
    var recordingWasEmpty by mutableStateOf(false)
        private set

    /** A brand-new clip recorded this session. */
    var recordedFile: File? by mutableStateOf(null)
        private set
    /** Storage path of an already-saved recording (when editing a word). */
    var existingAudioPath: String? by mutableStateOf(null)

    private var recorder: MediaRecorder? = null
    private var player: MediaPlayer? = null

    /** Whether there is any audio to play / save (new or previously saved). */
    val hasAudio: Boolean get() = recordedFile != null || existingAudioPath != null
    val hasNewRecording: Boolean get() = recordedFile != null

    fun configure(existingAudioPath: String?) {
        this.existingAudioPath = existingAudioPath
    }

    // MARK: Recording

    fun startRecording() {
        stopPlayback()
        recordingWasEmpty = false
        try {
            val file = File.createTempFile("retainic-rec-", ".m4a", context.cacheDir)
            @Suppress("DEPRECATION")
            val rec = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.S) MediaRecorder(context) else MediaRecorder()
            rec.setAudioSource(MediaRecorder.AudioSource.MIC)
            rec.setOutputFormat(MediaRecorder.OutputFormat.MPEG_4)
            rec.setAudioEncoder(MediaRecorder.AudioEncoder.AAC)
            rec.setAudioChannels(1)
            rec.setAudioSamplingRate(22_050)
            rec.setAudioEncodingBitRate(32_000)
            rec.setOutputFile(file.absolutePath)
            rec.prepare()
            rec.start()
            recorder = rec
            recordedFile = file
            isRecording = true
        } catch (e: Exception) {
            isRecording = false
        }
    }

    fun stopRecording() {
        try {
            recorder?.stop()
        } catch (_: Exception) {
        }
        recorder?.release()
        recorder = null
        isRecording = false

        val file = recordedFile
        // A valid AAC clip is well over 1 KB; treat tiny files as empty.
        if (file != null && file.length() < 1_000) {
            file.delete()
            recordedFile = null
            recordingWasEmpty = true
        }
    }

    /** Discards the current recording (new clip or reference to a saved one). */
    fun clear() {
        stopPlayback()
        if (isRecording) stopRecording()
        recordedFile?.delete()
        recordedFile = null
        existingAudioPath = null
    }

    // MARK: Playback

    fun play(scope: CoroutineScope) {
        val local = recordedFile
        val path = existingAudioPath
        when {
            local != null -> playLocal(local)
            path != null -> scope.launch { playRemote(path) }
        }
    }

    private fun playLocal(file: File) {
        try {
            val mp = MediaPlayer()
            mp.setDataSource(file.absolutePath)
            mp.setOnCompletionListener { isPlaying = false }
            mp.prepare()
            mp.start()
            player = mp
            isPlaying = true
        } catch (e: Exception) {
            isPlaying = false
        }
    }

    private suspend fun playRemote(path: String) {
        try {
            val data = VocabRepository.downloadAudioData(path)
            val tmp = withContext(Dispatchers.IO) {
                File.createTempFile("retainic-play-", ".m4a", context.cacheDir).apply { writeBytes(data) }
            }
            playLocal(tmp)
        } catch (e: Exception) {
            isPlaying = false
        }
    }

    fun stopPlayback() {
        player?.let {
            try { it.stop() } catch (_: Exception) {}
            it.release()
        }
        player = null
        isPlaying = false
    }
}

/**
 * Shared player for tapping a word's pronunciation outside the editor. Plays a
 * recorded clip (by Storage path) or a synthesized voice (TTS). Ported from
 * AudioPlaybackStore in AudioManager.swift.
 */
object AudioPlaybackStore {
    /** Storage path or "tts:"-prefixed key currently playing (for button state). */
    var playingPath: String? by mutableStateOf(null)
        private set

    private lateinit var appContext: Context
    private var player: MediaPlayer? = null
    private val cache = HashMap<String, ByteArray>()
    private val scope = CoroutineScope(Dispatchers.Main)

    private var tts: TextToSpeech? = null
    private var ttsReady = false

    fun init(context: Context) {
        appContext = context.applicationContext
        tts = TextToSpeech(appContext) { status ->
            ttsReady = status == TextToSpeech.SUCCESS
        }.apply {
            setOnUtteranceProgressListener(object : UtteranceProgressListener() {
                override fun onStart(utteranceId: String?) {}
                override fun onDone(utteranceId: String?) {
                    scope.launch { if (playingPath?.startsWith("tts:") == true) playingPath = null }
                }
                @Deprecated("Deprecated in Java")
                override fun onError(utteranceId: String?) {
                    scope.launch { if (playingPath?.startsWith("tts:") == true) playingPath = null }
                }
            })
        }
    }

    fun ttsKey(text: String): String = "tts:$text"

    fun toggle(path: String) {
        if (playingPath == path) stop() else scope.launch { play(path) }
    }

    /** Speaks [text] in [language], or stops if that utterance is already playing. */
    fun toggleSpeak(text: String, language: String) {
        val key = ttsKey(text)
        if (playingPath == key) {
            stop()
            return
        }
        stop()
        val engine = tts ?: return
        engine.language = Locale.forLanguageTag(bcp47(language))
        val slowdown = mapOf("en" to 0.95f, "ja" to 0.6f, "zh" to 0.75f)
        engine.setSpeechRate(slowdown[language] ?: 1.0f)
        playingPath = key
        engine.speak(text, TextToSpeech.QUEUE_FLUSH, null, key)
    }

    private fun bcp47(language: String): String = when (language) {
        "en" -> "en-US"
        "es" -> "es-ES"
        "zh" -> "zh-CN"
        "ja" -> "ja-JP"
        "ko" -> "ko-KR"
        else -> language
    }

    private suspend fun play(path: String) {
        stop()
        try {
            val data = cache[path] ?: withContext(Dispatchers.IO) {
                VocabRepository.downloadAudioData(path)
            }.also { cache[path] = it }
            val tmp = withContext(Dispatchers.IO) {
                File.createTempFile("retainic-store-", ".m4a", appContext.cacheDir).apply { writeBytes(data) }
            }
            val mp = MediaPlayer()
            mp.setDataSource(tmp.absolutePath)
            mp.setOnCompletionListener { playingPath = null }
            mp.prepare()
            mp.start()
            player = mp
            playingPath = path
        } catch (e: Exception) {
            playingPath = null
        }
    }

    fun stop() {
        player?.let {
            try { it.stop() } catch (_: Exception) {}
            it.release()
        }
        player = null
        tts?.let { if (it.isSpeaking) it.stop() }
        playingPath = null
    }
}
